apps/mymedicare_cb/models.py

```diff
@@ -57,43 +57,52 @@ def get_and_update_user(subject, mbi_hash, hicn_hash, first_name, last_name, ema
         user = User.objects.get(username=subject)
 
-        if user.crosswalk.user_hicn_hash != hicn_hash:
-            err_msg = "Found user's hicn did not match"
-            log_get_and_update_user(auth_flow_dict, "FAIL", user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, err_msg)
-            raise BBMyMedicareCallbackCrosswalkUpdateException(err_msg)
-
+        # fhir_id can not change for an existing user!
         if user.crosswalk.fhir_id != fhir_id:
-            err_msg = "Found user's fhir_id did not match"
-            log_get_and_update_user(auth_flow_dict, "FAIL", user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, err_msg)
-            raise BBMyMedicareCallbackCrosswalkUpdateException(err_msg)
+            mesg = "Found user's fhir_id did not match"
+            log_get_and_update_user(auth_flow_dict, "FAIL", user, fhir_id, mbi_hash, hicn_hash,
+                                    hash_lookup_type, crosswalk_updated, mesg)
+            raise BBMyMedicareCallbackCrosswalkUpdateException(mesg)
 
+        # Did the hicn change?
+        if user.crosswalk.user_hicn_hash != hicn_hash:
+            crosswalk_updated = crosswalk_updated + "H"
+
+        # Did the mbi change?
         if user.crosswalk.user_mbi_hash is not None:
             if user.crosswalk.user_mbi_hash != mbi_hash:
-                err_msg = "Found user's mbi did not match"
-                log_get_and_update_user(auth_flow_dict, "FAIL", user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, err_msg)
-                raise BBMyMedicareCallbackCrosswalkUpdateException(err_msg)
+                crosswalk_updated = crosswalk_updated + "M"
         else:
-            # Previously stored value was None/Null and mbi_hash != None, update just the mbi hash.
+            # Did the mbi change from previously stored None/Null value?
             if mbi_hash is not None:
-                err_msg = "UPDATE mbi_hash since previous value was NULL"
-                log_get_and_update_user(auth_flow_dict, "OK", user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, err_msg)
-                user.crosswalk.user_mbi_hash = mbi_hash
-                user.crosswalk.user_id_type = hash_lookup_type
-                user.crosswalk.save()
-
-        # Update hash type used for lookup, if it has changed from last match.
-        if user.crosswalk.user_id_type != hash_lookup_type:
-            err_msg = "UPDATE user_id_type as it has changed from the previous lookup value"
-            log_get_and_update_user(auth_flow_dict, "OK", user, fhir_id, mbi_hash, hicn_hash, hash_lookup_type, err_msg)
+                crosswalk_updated = crosswalk_updated + "N"
+
+        # Update Crosswalk if there are any allowed changes or hash_type used for lookup changed.
+        if user.crosswalk.user_id_type != hash_lookup_type or crosswalk_updated != "":
+            # Copy pre-save values to ArchivedCrosswalk instance
+            acw = ArchivedCrosswalk.objects.create(crosswalk=user.crosswalk,
```

Comment:
Think it'd be worth having a convenience function for this?
Something like:
`acw = ArchivedCrosswalk.objects.create(crosswalk)`
Where you just pass another crosswalk object into the function and it creates the archived crosswalk row from it?

Reply:
and put both updates in one TX demarcation:
 with transaction.atomic():
  .............................. update crosswalk and insert archived crosswalk.

Reply:
> Think it'd be worth having a convenience function for this?
> Something like:
> `acw = ArchivedCrosswalk.objects.create(crosswalk)`
> Where you just pass another crosswalk object into the function and it creates the archived crosswalk row from it?

@nbragdon I like this!  Just updating the code now.  Creating a `ArchivedCrosswalk.create()` method.

Reply:
> and put both updates in one TX demarcation:
> with transaction.atomic():
> .............................. update crosswalk and insert archived crosswalk.

@JFU-GIT Good catch and idea here!